man/systemd.exec.xml

```diff
@@ -1799,12 +1823,38 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
   <refsect1>
     <title>Environment variables in spawned processes</title>
 
-    <para>Processes started by the system are executed in a clean
-    environment in which select variables listed below are set. System
-    processes started by systemd do not inherit variables from PID 1,
-    but processes started by user systemd instances inherit all
-    environment variables from the user systemd instance.
-    </para>
+    <para>Processes started by the service manager are executed with an environment variable block assembled from
+    multiple sources. Processes started by the system service manager generally do not inherit environment variables
+    set for the service manager itself (but this may be altered via <varname>PassEnvironment=</varname>), but processes
+    started by the user service manager instances generally do inherit all environment variables set for the service
+    manager itself.</para>
+
+    <para>For each invoked process the list of environment variables set is compiled from the following sources:</para>
+
+    <itemizedlist>
+      <listitem><para>Variables globally configured for the service manager, using the
+      <varname>DefaultEnvironment=</varname> setting in
+      <citerefentry><refentrytitle>systemd-system.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>, the kernel command line option <varname>systemd.setenv=</varname> (see
+    <citerefentry><refentrytitle>systemd</refentrytitle><manvolnum>1</manvolnum></citerefentry>) or via
+      <command>systemctl set-environment</command> (see <citerefentry><refentrytitle>systemctl</refentrytitle><manvolnum>1</manvolnum></citerefentry>).</para></listitem>
+
+      <listitem><para>Variables defined by the service manager itself (see the list below)</para></listitem>
+
+      <listitem><para>Variables set in the service manager's own environment variable block (subject to <varname>PassEnvironment=</varname> for the system service manager)</para></listitem>
+
+      <listitem><para>Variables set via <varname>Environment=</varname> in the unit file</para></listitem>
+
+      <listitem><para>Variables read from files specified via <varname>EnvironmentFiles=</varname> in the unit file</para></listitem>
+
+      <listitem><para>Variables set by any PAM modules in case <varname>PAMName=</varname> is in effect, cf. <citerefentry project='man-pages'><refentrytitle>pam_env</refentrytitle><manvolnum>8</manvolnum></citerefentry></para></listitem>
+    </itemizedlist>
+
+    <para>If the same environment variables are set by multiple of these sources, the later source — according to the
+    order of the list above — wins. Note that as final step all variables listed in
+    <varname>UnsetEnvironment=</varname> are removed again from the compiled environment variable list, immediately
+    before it is passed to the executed process.</para>
```

Comment:
Say I have in my service file:
```
UnsetEnvironment=FOO
Environment=FOO=BAR
Environment=FOO=BAZ
```
Am I reading this correctly, that this will result in FOO being unset? Intuitively I would expect that last one wins, i.e. FOO=BAZ

Reply:
yes, it will be unset, and I added a big chunk of docs explaining the precise order in which these settings are applied

Reply:
Yeah, my point is that it's not quite intuitive that UnsetEnvironment doesn't respect the order and always wins.

On a more general note: everywhere else we have the scheme that an empty val unsets a setting, like ```ExecStart=```. Wouldn't it be more consistent if "Environment=FOO=" would unset FOO instead of setting it to an empty string? Besides consistency, this would have the benefit of not needing another keyword.

Reply:
> Wouldn't it be more consistent if Environment=FOO= would unset FOO instead of setting it to an empty string?

I don't think that the behaviour of `VAR=` should be changed in a backward incompatible way for the sake of consistency.

Maybe `VAR` could be used to unset what has been set with `VAR=`. For example,
```sh
Environment=A=1 B=2
Environment=A C=
```
would result in passing `B=2` and `C=`. Unfortunately, that still might break something.

Reply:
> Yeah, my point is that it's not quite intuitive that UnsetEnvironment doesn't respect the order and always wins.

Well, but that's not precisely new behaviour. It's not different from the relationship of PassEnvironment= to Environment=, or EnvironmentFiles= to Environment=: they all are applied in the order now documented and always have been, and not in the order they appear in the unit file. I think it's not overly surprising if UnsetEnvironment= works the same way, and is applied also not in the order the item is listed but in a fixed order.

Also, I think there's real benefit in being able to unset the env vars PID 1 itself sets, or the PAM modules set. That's why UnsetEnvironment= is applied as very last step of the logic, and not in the order it appears, because this would take this ability away...